src/Illuminate/Log/Context/ContextServiceProvider.php

```diff
@@ -0,0 +1,68 @@
+<?php
+
+namespace Illuminate\Log\Context;
+
+use Illuminate\Log\Context\Events\Dehydrating;
+use Illuminate\Log\Context\Events\Hydrated;
+use Illuminate\Queue\Events\JobProcessing;
+use Illuminate\Queue\Queue;
+use Illuminate\Support\ServiceProvider;
+
+class ContextServiceProvider extends ServiceProvider
+{
+    /**
+     * Register the service provider.
+     *
+     * @return void
+     */
+    public function register()
+    {
+        $this->app->singleton(Repository::class);
+    }
+
+    /**
+     * Boot the application services.
+     *
+     * @return void
+     */
+    public function boot()
+    {
+        Queue::createPayloadUsing(function ($c, $q, $payload) {
+            $context = $this->app[Repository::class];
+
+            $this->app['events']->dispatch(new Dehydrating($context));
+
+            if ($context->all() === [] && $context->allSecrets() === []) {
+                return $payload;
+            }
+
+            return [
+                ...$payload,
+                '__laravel_context' => [
+                    'data' => $context->all(),
+                    'secrets' => $context->allSecrets(),
+                ],
+            ];
+        });
+
+        $this->app['events']->listen(function (JobProcessing $event) {
+            $context = $this->app[Repository::class]->flush();
+
+            [
+                'data' => $data,
+                'secrets' => $secrets,
+            ] = $event->job->payload()['__laravel_context'] ?? ['data' => [], 'secrets' => []];
```

Comment:
I was unable to figure out what this was doing from GitHub because `$data` and `$secrets` didn't exist. I had to open an IDE to figure out this was setting those variables. I think this "smart tricky" will cause more struggle on folks diving into Laravel code than any benefit it might bring.

Reply:
I think it is fine.

Array destructuring has been available, to some degree, since PHP 4 via `list` with keyed support dropping in 7.1.

This syntax might be slightly newer, but I wouldn't consider it magic.